packages/cms-lib/userToken.js

```diff
@@ -0,0 +1,164 @@
+const open = require('open');
+const moment = require('moment');
+const {
+  promptUser,
+  USER_TOKEN_FLOW,
+  USER_TOKEN,
+} = require('@hubspot/cms-cli/lib/prompts');
+
+const { HubSpotAuthError } = require('@hubspot/api-auth-lib/Errors');
+const {
+  getPortalConfig,
+  updatePortalConfig,
+  updateDefaultPortal,
+  createEmptyConfigFile,
+  deleteEmptyConfigFile,
+} = require('./lib/config');
+const {
+  DEFAULT_HUBSPOT_CONFIG_YAML_FILE_NAME,
+  USER_TOKEN_AUTH_METHOD,
+} = require('./lib/constants');
+const { handleExit } = require('./lib/process');
+const { logger } = require('./logger');
+const { fetchAccessToken } = require('./api/localDevAuth');
+
+const refreshRequests = new Map();
+
+function getRefreshKeyForUserToken(userToken, expiration) {
+  return `${userToken}-${expiration || 'fresh'}`;
+}
+
+async function getAccessToken(userToken, env = 'QA') {
+  let response;
+  try {
+    response = await fetchAccessToken(userToken, env);
+    return {
+      portalId: response.hubId,
+      accessToken: response.oauthAccessToken,
+      expiresAt: moment(response.expiresAtMillis),
+      scopeGroups: response.scopeGroups,
+      encodedOauthRefreshToken: response.encodedOauthRefreshToken,
+    };
+  } catch (e) {
+    if (e.response) {
+      throw new HubSpotAuthError(
+        `Error while retrieving new access token: ${e.response.body.message}`
+      );
+    } else {
+      throw e;
+    }
+  }
+}
+
+async function refreshAccessToken(userToken, env = 'PROD') {
+  const { accessToken, expiresAt, portalId } = await getAccessToken(
+    userToken,
+    env
+  );
+  const config = getPortalConfig(portalId);
+
+  updatePortalConfig({
+    ...config,
+    portalId,
+    tokenInfo: {
+      accessToken,
+      expiresAt,
+    },
+  });
+
+  return accessToken;
+}
+
+async function getNewAccessToken(userToken, authTokenInfo, env) {
+  const key = getRefreshKeyForUserToken(
+    userToken,
+    authTokenInfo && authTokenInfo.expiresAt
+  );
+  if (refreshRequests.has(key)) {
+    return refreshRequests.get(key);
+  }
+  let accessToken;
+  try {
+    const refreshAccessPromise = refreshAccessToken(userToken, env);
+    if (key) {
+      refreshRequests.set(key, refreshAccessPromise);
+    }
+    accessToken = await refreshAccessPromise;
+  } catch (e) {
+    if (key) {
+      refreshRequests.delete(key);
+    }
+    throw e;
+  }
+  return accessToken;
+}
+
+async function accessTokenForUserToken(portalId) {
+  const { auth, userToken, env } = getPortalConfig(portalId);
+  const authTokenInfo = auth && auth.tokenInfo;
+  const authDataExists = authTokenInfo && auth.tokenInfo.accessToken;
+
+  if (
+    !authDataExists ||
+    moment()
+      .add(30, 'minutes')
+      .isAfter(moment(authTokenInfo.expiresAt))
+  ) {
+    return getNewAccessToken(
+      userToken,
+      authTokenInfo && authTokenInfo.expiresAt,
+      env
+    );
+  }
+
+  return auth.tokenInfo.accessToken;
+}
+
+/**
```

Comment:
Here down is new.

Reply:
It looks like there were changes above. Is that not correct?

Reply:
Yeah, this comment is outdated, you can ignore it. Will delete...